compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.codegen
+
+import com.intellij.util.PathUtil
+import com.intellij.util.SystemProperties
+import com.sun.jdi.ThreadReference
+import com.sun.jdi.VirtualMachine
+import com.sun.jdi.event.*
+import com.sun.jdi.request.EventRequest.SUSPEND_ALL
+import com.sun.jdi.request.EventRequest.SUSPEND_NONE
+import com.sun.jdi.request.StepRequest
+import com.sun.tools.jdi.SocketAttachingConnector
+import junit.framework.TestCase
+import org.jetbrains.kotlin.backend.common.output.OutputFile
+import org.jetbrains.kotlin.cli.common.output.writeAllTo
+import org.jetbrains.kotlin.codegen.state.GenerationState
+import org.jetbrains.kotlin.test.ConfigurationKind
+import org.jetbrains.kotlin.test.KotlinTestUtils
+import org.jetbrains.kotlin.utils.SmartList
+import java.io.File
+import java.io.IOException
+import java.lang.IllegalStateException
+import java.lang.reflect.Method
+import java.net.Socket
+import java.nio.charset.Charset
+import java.nio.file.Files
+import kotlin.properties.Delegates
+
+abstract class AbstractSteppingTest : CodegenTestCase() {
+
+    private companion object {
+        private const val DEBUG_ADDRESS = "127.0.0.1"
+        private const val MAIN_CLASS = "MainKt"
+        private const val BOX_METHOD = "box"
+        private const val LINENUMBER_PREFIX = "// LINENUMBER: "
+    }
+
+    data class RunJVMResult(val port: Int, val process: Process)
+
+    override fun doMultiFileTest(wholeFile: File, files: List<TestFile>) {
+        createEnvironmentWithMockJdkAndIdeaAnnotations(ConfigurationKind.ALL, *listOfNotNull(writeJavaFiles(files)).toTypedArray())
+
+        val lineNumbers = wholeFile
+            .readLines()
+            .filter { it.matches("$LINENUMBER_PREFIX.*".toRegex()) }
+            .map { line ->
+                line.drop(LINENUMBER_PREFIX.length)
+                    .trim()
+            }.joinToString(" ")
+
+        loadMultiFiles(files)
+        val classBuilderFactory = OriginCollectingClassBuilderFactory(ClassBuilderMode.FULL)
+        val generationState = GenerationUtils.compileFiles(myFiles.psiFiles, myEnvironment, classBuilderFactory)
+        classFileFactory = generationState.factory
+
+        val tempDirForTest = KotlinTestUtils.tmpDir("debuggerTest")
+
+        try {
+            val classesDir = File(tempDirForTest, "classes").apply {
+                classFileFactory.writeAllTo(this)
+            }
+
+            val (port, process) = startDebuggeeProcess(classesDir)
```

Comment:
Is it possible to avoid new process instantiation for each test? Maybe implement some simple server?
We have smth similar in TestProcessServer.kt to run test under JDK 1.6 (compiler requires JDK 1.8).
Actuall configuration is performed in compiler/tests-different-jdk/build.gradle.kts `codegenTest(target = 6, jvm = 6, jdk = "JDK_18")`

Reply:
Will that conflict with parallel test as paralleled tests should have different port numbers?

Reply:
Actually `codegenTest(target = 6, jvm = 6, jdk = "JDK_18")` tests are not parallelized. And seems  parallelization would not work properly with current approach of configuring and shuting down server with gradle. Seems it's best to instantiate server in test VM: in such case custom test runner would be required.

Reply:
I see, can you give me a pointer to how to customize test runner for this purpose?

Reply:
We have several custom runners that performs system preconfiguration in clinit, e.g. JUnit3WithIdeaConfigurationRunner. You can make similar approach starting server in clinit but's it's unclear how stop server cause shutdown hook wouldn't work as same JVM could be used for another tests. 
Maybe best to switch this tests to junit4 infrastructure with @BeforeClass @AfterClass routines.

Reply:
NB: seems problem with shutdown could be solved by RunListener.testRunFinished

Reply:
Sorry it took quiet a while for updating this PR, was packed by other stuff.

I've updated this test to make it only running process creation once during setup, let me know if the approach works. I modified TestProcessServer to meet my use case.